compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/symbols/BallerinaUnionTypeSymbol.java

```diff
@@ -70,7 +71,17 @@ public BallerinaUnionTypeSymbol(CompilerContext context, ModuleID moduleID, BFin
                 TypesFactory typesFactory = TypesFactory.getInstance(this.context);
 
                 for (BType memberType : ((BUnionType) this.getBType()).getMemberTypes()) {
-                    members.add(typesFactory.getTypeDescriptor(memberType));
+                    if (typesFactory.isTypeReference(memberType, memberType.tsymbol, false)
+                            || memberType.getKind() != TypeKind.FINITE) {
+                        members.add(typesFactory.getTypeDescriptor(memberType));
+                        continue;
+                    }
+
+                    BFiniteType finiteType = (BFiniteType) memberType;
+                    for (BLangExpression value : finiteType.getValueSpace()) {
+                        ModuleID moduleID = getModule().isPresent() ? getModule().get().id() : null;
+                        members.add(new BallerinaSingletonTypeSymbol(this.context, moduleID, value, value.getBType()));
```

Comment:
Not sure this is the correct way to do this. Doesn't `value.getBType()` return the broader type of that expression? IINM, we'll have to create individual `BFiniteType` types here for each of the member values.